src/pyedb/dotnet/edb_core/edb_data/primitives_data.py

```diff
@@ -160,13 +160,14 @@ def layer_name(self):
 
     @layer_name.setter
     def layer_name(self, val):
-        if isinstance(val, str) and val in list(self._core_stackup.layers.keys()):
-            lay = self._core_stackup.layers["TOP"]._edb_layer
-            if lay:
-                self.primitive_object.SetLayer(lay)
+        layer_list = list(self._core_stackup.layers.keys())
+        if isinstance(val, str) and val in layer_list:
+            layer = self._core_stackup.layers[val]._edb_layer
+            if layer:
+                self.primitive_object.SetLayer(layer)
             else:
-                raise AttributeError("Layer {} not found in layer".format(val))
-        elif isinstance(val, type(self._core_stackup.layers["TOP"])):
+                raise AttributeError("Layer {} not found.".format(val))
+        elif isinstance(val, type(self._core_stackup.layers[layer_list[0]])):
```

Comment:
Does it make sense to change the primitive object layer when providing something else than a string ?

Reply:
layer_list[0] is also a string. Problem is if in the stackup there is not a layer named "TOP", the function returns an error. So layer_list[0] calls the str name of the first layer in the stackup.

Reply:
Then i would advise to refactor the code to avoid calling `isinstance(val, str)` twice (once directly, the second time through `type(self._core_stackup.layers[layer_list[0]])`).

Reply:
isinstance(val, str) refers to input str name of layer: i.e. "TOP"
isinstance(val, type(self._core_stackup.layers[layer_list[0]])) refers to the actual object of layer: StackupLayerEdbClass.
So it works for both types.